// browser-sync/lib/default-config.js
module.exports = {
    debugInfo: true,
    injectFileTypes: ["css", "png", "jpg", "jpeg", "svg", "gif", "webp"],
    minify: true,
    logConnections: false,
    host: null,
    codeSync: true,
    timestamps: true,
    ghostMode: {
        clicks: true,
        links: false,
        forms: {
            submit: true,
            inputs: true,
            toggles: true
        },
        scroll: true,
        location: false
    },
    server: false,
    proxy: false,
    ports: {
        min: 3000,
        max: 4000
    },
    open: true,
    xip: false,
    hostnameSuffix: false,
    notify: true,
    devMode: false,
    debounce: 0,
    scrollProportionally: true,
    scrollThrottle: 0,
    reloadDelay: 0,
    injectChanges: true,
    startPath: null,
    excludedFileTypes: [
        "js",
        "css",
        "pdf",
        "map",
        "svg",
        "ico",
        "woff",
        "json",
        "eot",
        "ttf",
        "png",
        "jpg",
        "jpeg",
        "webp",
        "gif",
        "mp4",
        "mp3",
        "3gp",
        "ogg",
        "ogv",
        "webm",
        "m4a",
        "flv",
        "wmv",
        "avi",
        "swf",
        "scss"
    ]
};
